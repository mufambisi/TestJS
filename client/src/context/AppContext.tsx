import React, { createContext, useState, useContext } from "react";

export interface IContext {
  error?: Error;
}

export interface IAppContext {
  context: IContext;
  setContext(context: Partial<IContext>): void;
}

export const AppContext = createContext<IAppContext>({
  context: {},
  setContext: /* istanbul ignore next */ () => undefined,
});

export const mergeContext = ({ setContext, context }: IAppContext) => (
  nextContext: Partial<IContext>
): void =>
  setContext({
    ...context,
    ...nextContext,
  });

export const useAppContext = (): IAppContext =>
  useContext<IAppContext>(AppContext);

export const AppContextProvider: React.FC = ({ children }) => {
  const [context, setContext] = useState<IContext>({});

  return (
    <AppContext.Provider
      value={{
        context,
        setContext: mergeContext({ setContext, context }),
      }}
    >
      {children}
    </AppContext.Provider>
  );
};

const AppContextConsumer = AppContext.Consumer;
export { AppContextConsumer };
