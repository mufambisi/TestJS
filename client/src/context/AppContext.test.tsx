import React from "react";
import { render } from "@testing-library/react";

import { AppContextProvider, mergeContext } from "./AppContext";

describe("<AppContext />", () => {
  it("should render without error", () => {
    const { container } = render(
      <AppContextProvider>Hello world</AppContextProvider>
    );

    expect(container).toMatchSnapshot();
  });
});

describe("mergeContext()", () => {
  it("should call setContext() with merged context", () => {
    const setContext = jest.fn();

    const status = "OK";
    const error = new Error("Error");
    const context = {
      status,
    } as any;

    mergeContext({ setContext, context })({ error });

    expect(setContext).toHaveBeenCalledWith({
      error,
      status,
    });
  });
});
