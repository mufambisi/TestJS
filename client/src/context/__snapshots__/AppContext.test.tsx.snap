// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppContext /> should render without error 1`] = `
<div>
  Hello world
</div>
`;
